src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/CodeMetricsAnalyzer.cs

```diff
@@ -0,0 +1,485 @@
+                                                                     helpLinkUri: "https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1506-avoid-excessive-class-coupling",
+                                                                     customTags: WellKnownDiagnosticTags.Telemetry);
+
+        internal static DiagnosticDescriptor InvalidEntryInCodeMetricsConfigFileRule = new DiagnosticDescriptor(CA1508RuleId,
+                                                                     s_localizableTitleCA1508,
+                                                                     s_localizableMessageCA1508,
+                                                                     DiagnosticCategory.Maintainability,
+                                                                     DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                     isEnabledByDefault: false,
+                                                                     description: s_localizableDescriptionCA1508,
+                                                                     helpLinkUri: null, // TODO: Add help link
+                                                                     customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(CA1501Rule, CA1502Rule, CA1505Rule, CA1506Rule, InvalidEntryInCodeMetricsConfigFileRule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationAction(compilationContext =>
+            {
+                // Try read the additional file containing the code metrics configuration.
+                if (!TryGetRuleIdToThresholdMap(
+                        compilationContext.Options.AdditionalFiles,
+                        compilationContext.CancellationToken,
+                        out AdditionalText additionalTextOpt,
+                        out ImmutableDictionary<string, List<(SymbolKind?, uint)>> ruleIdToThresholdMap,
+                        out List<Diagnostic> invalidFileDiagnostics) &&
+                    invalidFileDiagnostics != null)
+                {
+                    // Report any invalid additional file diagnostics.
+                    foreach (var diagnostic in invalidFileDiagnostics)
+                    {
+                        compilationContext.ReportDiagnostic(diagnostic);
+                    }
+                }
+
+                // Compute code metrics.
+                var computeTask = Task.Run(() => CodeAnalysisMetricData.ComputeAsync(compilationContext.Compilation, compilationContext.CancellationToken));
```

Comment:
is there a reason who run it on Task.Run and then wait on that? rather than directly wait on task returned by ComputAsync?